test/shared/src/main/scala/zio/test/Gen.scala

```diff
@@ -18,47 +18,99 @@ package zio.test
 
 import zio.{ UIO, ZIO }
 import zio.random._
-import zio.stream.ZStream
+import zio.stream.{ Stream, ZStream }
 
 /**
  * A `Gen[R, A]` represents a generator of values of type `A`, which requires
  * an environment `R`. Generators may be random or deterministic.
  */
 case class Gen[-R, +A](sample: ZStream[R, Nothing, Sample[R, A]]) { self =>
-  final def <*>[R1 <: R, B](that: Gen[R1, B]): Gen[R1, (A, B)] = self zip that
 
-  final def flatMap[R1 <: R, B](f: A => Gen[R1, B]): Gen[R1, B] =
-    Gen(sample.flatMap(sample => f(sample.value).sample))
+  final def <*>[R1 <: R, B](that: Gen[R1, B]): Gen[R1, (A, B)] =
+    self zip that
 
-  final def map[B](f: A => B): Gen[R, B] = Gen(sample.map(_.map(f)))
+  final def filter(f: A => Boolean): Gen[R, A] =
+    flatMap(a => if (f(a)) Gen.const(a) else Gen.empty)
 
-  final def zip[R1 <: R, B](that: Gen[R1, B]): Gen[R1, (A, B)] = self.flatMap(a => that.map(b => (a, b)))
+  final def flatMap[R1 <: R, B](f: A => Gen[R1, B]): Gen[R1, B] = Gen {
+    sample.flatMap { a =>
+      f(a.value).sample.map { b =>
+        val rest = a.shrink.flatMap(a => f(a).sample).flatMap(sample => ZStream(sample.value) ++ sample.shrink)
```

Comment:
I wonder what the most useful way to define this is...

Reply:
I think this is one of the key issues. I think, though I could be wrong, that this is the only lawful way to define it for potentially finite streams. The challenge with the stream based flatMap (at least when the stream is not just a single element that generates a random effect) is that we have to essentially traverse all the children of the first possibility before we can explore other possibilities. So like:

```scala
for {
  x <- Stream(1, 2, 3)
  y <- Stream(4, 5, 6)
} yield (x, y)
// Stream((1, 4), (1, 5), (1, 6), (2, 4), ...)
```

We get all the `1` possibilities before we see other possibilities whereas a lot of the time we want a more balanced distribution which is where combinators like `union` come in to get a more balanced distribution.

The other thing we could explore is the infinite stream instance where `unit` is repeating the same value and `flatMap` is taking the diagonal but I'm not sure that is where we want to go.